### PR TITLE
fix(lsp): stop diagnostics flickering

### PR DESCRIPTION
**Pull request by dsherret**
https://github.com/denoland/deno/assets/1609021/4d00df19-6747-4f16-9c6d-e2e95fad9755

Closes https://github.com/denoland/vscode_deno/issues/835

### CODE REVIEW

File: cli/lsp/diagnostics.rs [hunk 1]
```diff
@@ -351,16 +453,18 @@ impl DiagnosticsServer {
...
-                    &config,
-                    token.clone(),
-                  )
-                  .await;
+                  let diagnostics = spawn_blocking({
+                    let token = token.clone();
+                    move || generate_deno_diagnostics(&snapshot, &config, token)
+                  })
+                  .await
+                  .unwrap();
```
Comment (dsherret) [inline cli/lsp/diagnostics.rs:461]: Also gets the deno and lint diagnostics in blocking tasks, which should speed things up.

File: cli/lsp/diagnostics.rs [hunk 2]
```diff
@@ -54,61 +56,145 @@ pub struct DiagnosticServerUpdateMessage {
...
+      *source_diagnostics = record.versioned;
+
+      // DO NOT filter these by version. We want to display even out
+      // of date diagnostics in order to prevent flickering. The user's
+      // lsp client will eventually catch up.
+      let all_specifier_diagnostics = diagnostics_by_source
+        .values()
+        .flat_map(|d| &d.diagnostics)
+        .cloned()
+        .collect::<Vec<_>>();
```
Comment (dsherret) [inline cli/lsp/diagnostics.rs:138]: The main fix is here. We store the diagnostics by specifier then source (deno, lint, ts) and then don't filter on version.

File: cli/lsp/diagnostics.rs [hunk 3]
```diff
@@ -66,62 +68,133 @@ struct VersionedDiagnostics {
...
-type DiagnosticMap = HashMap<ModuleSpecifier, VersionedDiagnostics>;
-type DiagnosticsByVersionMap = HashMap<Option<i32>, Vec<lsp::Diagnostic>>;
 
-#[derive(Clone)]
+#[derive(Debug, Hash, PartialEq, Eq, Copy, Clone)]
+pub enum DiagnosticSource {
+  Deno,
+  Lint,
+  Ts,
+}
```
Comment (bartlomieju) [inline cli/lsp/diagnostics.rs:77]: Very nice 👍

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks for this fix!

**Pull request merged**